pybpodapi/bpod/bpod_base.py

```diff
@@ -560,39 +671,89 @@ def __update_timestamps(self, sma, state_change_indexes):
+                timestamps = [
+                    float(t) * self._hardware.times_scale_factor for t in timestamps]
+        if self._emulator.enabled or not self.hardware.live_timestamps:
             # update the timestamps of the events #############################################################
             for event, timestamp in zip(current_trial.events_occurrences, timestamps):
                 event.host_timestamp = timestamp
-                e = EventResume(event.event_id, event.event_name, host_timestamp=timestamp)
+                e = EventResume(event.event_id, event.event_name,
+                                host_timestamp=timestamp)
                 self.session += e
             ###################################################################################################
-
         current_trial.event_timestamps = timestamps
-        current_trial.state_timestamps += [timestamps[i] for i in state_change_indexes]
+        current_trial.state_timestamps += [timestamps[i]
+                                           for i in state_change_indexes]
         current_trial.state_timestamps += timestamps[-1:]
 
+    def manual_override(self, channel_type, channel_name, channel_number, value):
+        """
+        Manually override a Bpod channel
+
+        :param ChannelType channel_type: channel type input or output
+        :param ChannelName channel_name: channel name like PWM, Valve, etc.
+        :param channel_number:
+        :param int value: value to write on channel
+        """
+        if channel_type == ChannelType.INPUT:
+            input_channel_name = channel_name + str(channel_number)
+            try:
+                channel_number = self.hardware.channels.input_channel_names.index(
+                    input_channel_name)
+                self.trigger_input(channel_number, value)
+            except:
+                raise BpodErrorException(
+                    'Error using manual_override: {name} is not a valid channel name.'.format(name=channel_name))
+
+        elif channel_type == ChannelType.OUTPUT:
+            if channel_name == 'SoftCode':
+                self.trigger_softcode(value)
+            elif channel_name == 'Serial':
+                self.trigger_serial(channel_number, value)
+            else:
+                output_channel_name = channel_name + \
+                    str(channel_number)
+                try:
+                    channel_number = self.hardware.channels.output_channel_names.index(
+                        output_channel_name)
+                    self.trigger_output(channel_number, value)
+                except:
+                    raise BpodErrorException('Error using manual_override: {name} is not a valid channel name.'.format(
+                        name=output_channel_name))
+        else:
+            raise BpodErrorException(
+                'Error using manualOverride: first argument must be "Input" or "Output".')
+
```

Comment:
> Is there a reason for moving this function from the bpod_com_protocol module to here? I failed to spot a necessity for it.

Since I add emulator logic, the function no longer contains only com logic. Hence, the move.

Reply:
But nothing the code there is an emulator code, it's mainly calling base class functions. In such case, it might more sense to leave it in the child class.
Also, some of the changes here are unrelated, e.g: moving code inside the try/catch block.

Reply:
The trigger_* functions that this function calls check whether the emulator mode is enabled and dispatch an action accordingly. I made the move since it no longer **only** calls 'com' stuff, but also emulator stuff (through the trigger functions).